File: copymove/src/main/scala/farjs/copymove/CopyItemsStats.scala
package farjs.copymove

import farjs.ui.popup.{StatusPopup, StatusPopupProps}
import farjs.ui.task.{Task, TaskAction}
import scommons.react._
import scommons.react.hooks._

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.scalajs.js
import scala.util.{Failure, Success}

object CopyItemsStats extends FunctionComponent[CopyItemsStatsProps] {

  private[copymove] var statusPopupComp: ReactClass = StatusPopup

  protected def render(compProps: Props): ReactElement = {
    val (currDir, setCurrDir) = useState("")
    val inProgress = useRef(false)
    val props = compProps.plain

    def scanDir(): Unit = {
      val parent = props.fromPath
      var filesSize = 0.0
      val resultF = props.items.foldLeft(Future.successful(true)) { (resF, currItem) =>
        resF.flatMap {
          case true if currItem.isDir && inProgress.current =>
            setCurrDir(currItem.name)
            props.actions.scanDirs(parent, js.Array(currItem), onNextDir = { (_, items) =>
              filesSize += items.foldLeft(0d) { case (res, i) =>
                res + (if (i.isDir) 0d else i.size)
              }
              inProgress.current
            }).toFuture
          case true if !currItem.isDir =>
            filesSize += currItem.size
            Future.successful(true)
          case res => Future.successful(res)
        }
      }
      resultF.onComplete {
        case Success(false) => // already cancelled
        case Success(true) =>
          props.onDone(filesSize)
        case Failure(_) =>
          props.onCancel()
          props.dispatch(TaskAction(Task(s"${props.title} dir scan", resultF)))
      }
    }

    useLayoutEffect({ () =>
      // start scan
      inProgress.current = true
      scanDir()
      
      val cleanup: js.Function0[Unit] = { () =>
        // stop scan
        inProgress.current = false
      }
      cleanup
    }, Nil)

    <(statusPopupComp)(^.plain := StatusPopupProps(
      text = s"Calculating total size\n$currDir",
      title = props.title,
      onClose = props.onCancel
    ))()
  }
}

File: copymove/CopyItemsStats.mjs
/**
 * @import { Dispatch } from "@farjs/filelist/FileListData.mjs"
 * @import { FileListItem } from "@farjs/filelist/api/FileListItem.mjs"
 */
import React, { useLayoutEffect, useRef, useState } from "react";
import Task from "@farjs/ui/task/Task.mjs";
import TaskAction from "@farjs/ui/task/TaskAction.mjs";
import StatusPopup from "@farjs/ui/popup/StatusPopup.mjs";
import FileListActions from "@farjs/filelist/FileListActions.mjs";

const h = React.createElement;

/**
 * @typedef {{
 *  readonly dispatch: Dispatch;
 *  readonly actions: FileListActions;
 *  readonly fromPath: string;
 *  readonly items: readonly FileListItem[];
 *  readonly title: string;
 *  onDone(total: number): void;
 *  onCancel(): void;
 * }} CopyItemsStatsProps
 */

/**
 * @param {CopyItemsStatsProps} props
 */
const CopyItemsStats = (props) => {
  const { statusPopupComp } = CopyItemsStats;

  const [currDir, setCurrDir] = useState("");
  const inProgress = useRef(false);

  const scanDir = () => {
    const parent = props.fromPath;
    let filesSize = 0;

    const resultP = props.items.reduce(async (resP, currItem) => {
      const res = await resP;
      if (res && currItem.isDir && inProgress.current) {
        setCurrDir(currItem.name);
        return props.actions.scanDirs(parent, [currItem], (_, items) => {
          filesSize += items.reduce((res, i) => {
            return res + (i.isDir ? 0 : i.size);
          }, 0);
          return inProgress.current;
        });
      }
      if (res && !currItem.isDir) {
        filesSize += currItem.size;
        return true;
      }
      return res;
    }, Promise.resolve(true));

    resultP.then(
      (res) => {
        if (res) {
          props.onDone(filesSize);
        }
        //else: already cancelled
      },
      () => {
        props.onCancel();
        props.dispatch(TaskAction(Task(`${props.title} dir scan`, resultP)));
      }
    );
  };

  useLayoutEffect(() => {
    // start scan
    inProgress.current = true;
    scanDir();

    const cleanup = () => {
      // stop scan
      inProgress.current = false;
    };
    return cleanup;
  }, []);

  return h(statusPopupComp, {
    text: `Calculating total size\n${currDir}`,
    title: props.title,
    onClose: props.onCancel,
  });
};

CopyItemsStats.displayName = "CopyItemsStats";
CopyItemsStats.statusPopupComp = StatusPopup;

export default CopyItemsStats;
